### ISSUE TITLE
tf.keras.layers.Dropout does not accept noise_shape with None-dimension

### ISSUE THREAD

**Issue by Callidior** [tensorflow/tensorflow#30946]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux openSUSE 42.2
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary (via conda)
- TensorFlow version (use command below): 1.14.0
- Python version: 3.7.3
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version: CUDA 9.0 / cuDNN 7.6.0
- GPU model and memory: GeForce GTX 1080 Ti (10.92 GB)

**Describe the current behavior**

The Dropout layer accepts a `noise_shape` argument. If set, this shape has to include the batch size, which is usually not known at model construction time. Thus, this argument is actually only useful if `None` can be specified for any axis and is then expanded to the actual size of that axis in the input tensor passed to the layer.

This is currently perfectly possible in standard `keras` but raises an error in `tf.keras` (see below).

**Describe the expected behavior**

It should be possible to include `None`-sized axes in the `noise_shape` argument to `tf.keras.layers.Dropout`, which are expanded at run-time to the actual size of that axis in the input tensor of the layer.

**Code to reproduce the issue**

```python
from tensorflow import keras

inp_ = keras.layers.Input((2,3,4))
dropout = keras.layers.Dropout(0.5, noise_shape=(None,1,1,1))
x = dropout(inp_, training=True)
dropfun = keras.backend.function([inp_], [x])
```

This code works completely fine if the first line is replaced with `import keras`.

**Other info / logs**

Traceback:
```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-1601abc84b49> in <module>
      1 inp_ = keras.layers.Input((2,3,4))
      2 dropout = keras.layers.Dropout(0.5, noise_shape=(None,1,1,1))
----> 3 x = dropout(inp_, training=True)
      4 dropfun = keras.backend.function([inp_], [x])

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/keras/engine/base_layer.py in __call__(self, inputs, *args, **kwargs)
    632                     outputs = base_layer_utils.mark_as_return(outputs, acd)
    633                 else:
--> 634                   outputs = call_fn(inputs, *args, **kwargs)
    635 
    636             except TypeError as e:

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/keras/layers/core.py in call(self, inputs, training)
    160     output = tf_utils.smart_cond(training,
    161                                  dropped_inputs,
--> 162                                  lambda: array_ops.identity(inputs))
    163     return output
    164 

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/keras/utils/tf_utils.py in smart_cond(pred, true_fn, false_fn, name)
     56         pred, true_fn=true_fn, false_fn=false_fn, name=name)
     57   return smart_module.smart_cond(
---> 58       pred, true_fn=true_fn, false_fn=false_fn, name=name)
     59 
     60 

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/framework/smart_cond.py in smart_cond(pred, true_fn, false_fn, name)
     52   if pred_value is not None:
     53     if pred_value:
---> 54       return true_fn()
     55     else:
     56       return false_fn()

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/keras/layers/core.py in dropped_inputs()
    156           noise_shape=self._get_noise_shape(inputs),
    157           seed=self.seed,
--> 158           rate=self.rate)
    159 
    160     output = tf_utils.smart_cond(training,

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/util/deprecation.py in new_func(*args, **kwargs)
    505                 'in a future version' if date is None else ('after %s' % date),
    506                 instructions)
--> 507       return func(*args, **kwargs)
    508 
    509     doc = _add_deprecated_arg_notice_to_docstring(

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py in dropout(x, keep_prob, noise_shape, seed, name, rate)
   4168     raise ValueError("You must provide a rate to dropout.")
   4169 
-> 4170   return dropout_v2(x, rate, noise_shape=noise_shape, seed=seed, name=name)
   4171 
   4172 

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/ops/nn_ops.py in dropout_v2(x, rate, noise_shape, seed, name)
   4247     # and subtract 1.0.
   4248     random_tensor = random_ops.random_uniform(
-> 4249         noise_shape, seed=seed, dtype=x.dtype)
   4250     keep_prob = 1 - rate
   4251     scale = 1 / keep_prob

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/ops/random_ops.py in random_uniform(shape, minval, maxval, dtype, seed, name)
    237     maxval = 1
    238   with ops.name_scope(name, "random_uniform", [shape, minval, maxval]) as name:
--> 239     shape = _ShapeTensor(shape)
    240     minval = ops.convert_to_tensor(minval, dtype=dtype, name="min")
    241     maxval = ops.convert_to_tensor(maxval, dtype=dtype, name="max")

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/ops/random_ops.py in _ShapeTensor(shape)
     42   else:
     43     dtype = None
---> 44   return ops.convert_to_tensor(shape, dtype=dtype, name="shape")
     45 
     46 

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in convert_to_tensor(value, dtype, name, preferred_dtype, dtype_hint)
   1085   preferred_dtype = deprecation.deprecated_argument_lookup(
   1086       "dtype_hint", dtype_hint, "preferred_dtype", preferred_dtype)
-> 1087   return convert_to_tensor_v2(value, dtype, preferred_dtype, name)
   1088 
   1089 

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in convert_to_tensor_v2(value, dtype, dtype_hint, name)
   1143       name=name,
   1144       preferred_dtype=dtype_hint,
-> 1145       as_ref=False)
   1146 
   1147 

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/framework/ops.py in internal_convert_to_tensor(value, dtype, name, as_ref, preferred_dtype, ctx, accept_symbolic_tensors, accept_composite_tensors)
   1222 
   1223     if ret is None:
-> 1224       ret = conversion_func(value, dtype=dtype, name=name, as_ref=as_ref)
   1225 
   1226     if ret is NotImplemented:

~/lib/anaconda3/envs/tf114/lib/python3.7/site-packages/tensorflow/python/framework/constant_op.py in _tensor_shape_tensor_conversion_function(s, dtype, name, as_ref)
    324   if not s.is_fully_defined():
    325     raise ValueError(
--> 326         "Cannot convert a partially known TensorShape to a Tensor: %s" % s)
    327   s_list = s.as_list()
    328   int64_value = 0

ValueError: Cannot convert a partially known TensorShape to a Tensor: (?, 1, 1, 1)
```

The bug is probably due to the `tf.keras.layers.Dropout._get_noise_shape` function, which explicitly performs the `None`-expansion in [standard keras](https://github.com/keras-team/keras/blob/master/keras/layers/core.py#L110).

**Comment by yongtang** [comment]
Looking into the code history, I think this issue might have been fixed in: ab2644e

**Comment by Callidior** [comment]
Looks good. Thanks for pointing me to this and sorry for redundant reporting.
